rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,512 @@
+to corruption if multiple independent updates occur simultaneously.
+
+Unfortunately, AWS S3 does not provide any built-in methods for preventing this
+sort of issue.
+
+It also apparently lacks the primitives that would allow us to layer one over S3:
+
+* There is no way to push an object to a bucket only if an object with the same
+  key does not already exist, which means we can't really implement a race free
+  lockfile-like protocol
+* There are no compare-and-swap style put primitives
+* While versioning _is_ supported, there is no way to say "only overwrite object
+  _X_ if its current version is _Y_".
+
+Our initial approach may have to be relying on the build serialisation tools
+provided by Drone.
+
+#### Build Security issues
+
+This process will require that the `teleport-plugins` promotion process have
+access to new secrets & resources:
+
+##### Staging Resources
+
+* Staging Registry Bucket
+* Staging Registry CloudFront
+* Staging Registry logs bucket
+* Staging IAM User with read/write access to Staging Registry Bucket
+  * Modify permission required on `versions` file
+* Staging signing key
+
+##### Production Resources
+
+* Production Registry Bucket
+* Production Registry CloudFront
+* Production Registry logs bucket
+* Production IAM User with read/write access to Production Registry Bucket
+  * Modify permission required on `versions` file
+* Production signing key
+
+_**Note:**_ We currently expect to use the RPM signing key as the Terraform
+production signing key as it already exists and is already used for a similar
+purpose.
+
+That said, there is no technical reason why it _must_ be the RPM signing
+key. As long as the private key used to sign the package agrees with the a
+public key exposed via the registry, the signature should be deemed valid
+by Terraform.
+
+_**Also note**_ that we can rotate the key used for signing _new_ additions to
+the registry without having to re-sign the entire registry, meaning that we
+can get started using the existing RPM signing key and cheaply rotate it when
+a new key management structure is in place.
+
+### Testing
+
+The system will be tested by promoting to a "staging" environment. The
+staging environment will be selected during the Drone promotion process
+and will use a dummy bucket and signing key.
+
+## Open Questions
+
+### How should we name the provider?
+
+We have three names we need to pick, and these will change how users will
+include the Teleport provider in their terraform scripts:
+
+1) the registry host; current options are:
+   * `goteleport.com`
+   * `terraform.releases.teleport.dev` (i.e. matching existing distribution site naming scheme)
+   * `terraform.goteleport.com` (something in-between)
+   * ... something else entirely?
+2) the provider namespace; options are `gravitational` or `teleport`,
+3) the provider name; the obvious choice is `teleport`, but mentioned for
+   completeness
```

Comment:
I vote for the shorter domain (easier to remember), and "gravitational" namespace in case we decide to add a provider for something other than Teleport itself.

Reply:
We have helm charts and other assets are already on teleport.dev (.dev is operated by google).
https://charts.releases.teleport.dev/[](https://charts.releases.teleport.dev/)

Why not [teleport.dev/terraform](http://teleport.dev/terraform)
Either way, it can be goteleport.com, but in this case we should move helm charts and other assets to goteleport.com.I think all our assets should be distributed from one domain.
Do you agree?

Reply:
Unfortunately `teleport.dev/terraform` wouldn't match the requirements of the Terraform discovery protocol, but we could certainly do something like `terraform.teleport.dev/gravitational/teleport`

Reply:
I'd encourage us to stick with the same `terraform.releases.teleport.dev` style as we have for the current repos for the sake of consistency. The full name `terraform.releases.teleport.dev/gravitational/teleport` is long and a little redundant, but I think that for something that users are going to copy/paste once and use for a longer period, it's not going to be a huge issue.